### PR TITLE
Use a longer timeout when powering back the Thunderbolt device

### PR DESCRIPTION
**Pull request by hughsie**
It seems 2s is not long enough on some hardware. We also get a much better
error message if the device fails to come back online.

### CODE REVIEW

**Comment by gicmo** [comment]
lgtm, fwiw ;)

**Comment by superm1** [comment]
Is this patch a guess based on TBT failures that have been reported to LVFS, or something seen first hand?  I was suspecting something like this too, but I was also hoping to catch one of those failures in the wild with more metadata before jumping to a conclusion that it was the removal of force power too quickly.

**Comment by gicmo** [comment]
@superm1 we (@hughsie) did a quick analysis of a reported thunderbolt update failures, where the error message is `could not find thunderbolt device at <path>`, where `<path>` is indicative of the device being the host controller. It seems that we should only get that error if we ignored the remove event, which we only do for force-powerable devices. The kernel version is on these reports is `4.15` which supports force-powering. So the timeout being it would be my guess.
Additionally I really like the better error message we get with that patch (i.e. "could not force power" vs "could not find device").

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c [hunk 1]
```diff
@@ -240,11 +240,23 @@ fu_plugin_update_prepare (FuPlugin *plugin,
 		return FALSE;
 
 	data->needs_forcepower = TRUE;
+
 	/* wait for the device to come back onto the bus */
 	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
-	g_usleep (TBT_NEW_DEVICE_TIMEOUT * G_USEC_PER_SEC);
+	for (guint i = 0; i < 10; i++) {
```
Comment (superm1) [inline plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: So this is effectively waiting 4 minutes (25s * 10 tries)?  That seems pretty long to me.  If it doesn't come back in the first minute it's probably not going to I would say.

**Comment by superm1** [comment]
@gicmo thanks for the clarification.  It would be good to also report in the metadata that it was force powered (right now it's just a guess based on 4.15+).  Would you be able to add that too?

**Comment by ybernat** [comment]
My 2 cents:
1. I think the general approach is better, checking that the force power actually brought the controller up instead of just waiting for it.
2. The longest wait I ever heard as  a general recommendation for force-power is 10 seconds. You may want to double it just-in-case, but 4 minutes sounds like a way too long wait indeed.

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c [hunk 2]
```diff
@@ -240,11 +240,23 @@ fu_plugin_update_prepare (FuPlugin *plugin,
...
+		g_autoptr(GUdevDevice) udevice_tmp = NULL;
+		g_usleep (250000);
```
Comment (ybernat) [inline plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: What's is the guideline here about defines vs. hard-coded numbers?

**Comment by superm1** [comment]
>The longest wait I ever heard as a general recommendation for force-power is 10 seconds. You may want to double it just-in-case, but 4 minutes sounds like a way too long wait indeed.

Yeah maybe instead increase `TBT_NEW_DEVICE_TIMEOUT `to 10s then and lower the number of passes (say maybe 2 or 3).

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c (hunk 2 above)
Comment (superm1) [reply plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: I think I would also rather see TBT_NEW_DEVICE_TIMEOUT increased than a hard-coded number.

**Comment by ybernat** [comment]
> Yeah maybe instead increase TBT_NEW_DEVICE_TIMEOUTto 10s then and lower the number of passes (say maybe 2 or 3).

(Or vice versa, keeping the timeout to 2s and just let it retry a few times, e.g. 10, to exit the retry loop sooner if the controller is up quickly)

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c (hunk 1 above)
Comment (hughsie) [reply plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: 10x250000 = 10x250ms = 2.5s

**Comment by hughsie** [comment]
Okay, what about this version?

**Pull request merged**